napi: Remove napi_set_function_name since not all engines can rename functions

boingoing opened the pull request:
Add an optional name parameter to napi_create_function.

Also change V8LocalValueFromJsPropertyName to V8LocalStringFromJsPropertyName instead since these are always string values.

boingoing committed:
Remove napi_set_function_name since not all engines can rename functions

Add an optional name parameter to napi_create_function.

Also change V8LocalValueFromJsPropertyName to V8LocalStringFromJsPropertyName instead since these are always string values.

mhdawson reviewed (changes requested):
ignore wrong issue, will re-review

mhdawson commented:
Looking at the error message seems like a typo in  reintepret_cast

Did you run make test after the changes ?

mhdawson reviewed (approved):
LGTM, but should probably land after the one fixing up the cpplint errors.

CI run: https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-commit-linux-abi/18/

mhdawson commented:
```bash

Seems to have failed CI, gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
make[2]: Entering directory `/home/iojs/build/workspace/node-test-commit-linux-abi/nodes/ubuntu1404-64/test/addons/02_object_factory/build'
  CXX(target) Release/obj.target/addon/addon.o
../binding.cc: In function 'void CreateFunction(napi_env, napi_callback_info)':
../binding.cc:26:48: error: 'napi_set_function_name' was not declared in this scope
   status = napi_set_function_name(env, fn, name);
                                                ^
../binding.cc: In function 'void Init(napi_env, napi_value, napi_value)':
../binding.cc:35:63: warning: missing initializer for member 'napi_property_descriptor::getter' [-Wmissing-field-initializers]
   napi_property_descriptor desc = { "exports", CreateFunction };
                                                               ^
../binding.cc:35:63: warning: missing initializer for member 'napi_property_descriptor::setter' [-Wmissing-field-initializers]
../binding.cc:35:63: warning: missing initializer for member 'napi_property_descriptor::value' [-Wmissing-field-initializers]
../binding.cc:35:63: warning: missing initializer for member 'napi_property_descriptor::attributes' [-Wmissing-field-initializers]
../binding.cc:35:63: warning: missing initializer for member 'napi_property_descriptor::data' [-Wmissing-field-initializers]
make[2]: *** [Release/obj.target/binding/binding.o] Error 1
make[2]: Leaving directory `/home/iojs/build/workspace/node-test-commit-linux-abi/nodes/ubuntu1404-64/test/addons-abi/5_function_factory/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/home/iojs/build/workspace/node-test-commit-linux-abi/nodes/ubuntu1404-64/deps/npm/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:192:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
gyp ERR! System Linux 3.13.0-71-generic
gyp ERR! command "/home/iojs/build/workspace/node-test-commit-linux-abi/nodes/ubuntu1404-64/out/Release/node" "/home/iojs/build/workspace/node-test-commit-linux-abi/nodes/ubuntu1404-64/deps/npm/node_modules/node-gyp/bin/node-gyp" "rebuild" "--python=/usr/bin/python" "--directory=/home/iojs/build/workspace/node-test-commit-linux-abi/nodes/ubuntu1404-64/test/addons-abi/5_function_factory/" "--nodedir=/home/iojs/build/workspace/node-test-commit-linux-abi/nodes/ubuntu1404-64"
gyp ERR! cwd /home/iojs/build/workspace/node-test-commit-linux-abi/nodes/ubuntu1404-64/test/addons-abi/5_function_factory
gyp ERR! node -v v8.0.0-pre
gyp ERR! node-gyp -v v3.5.0
gyp ERR! not ok 
make[1]: *** [test/addons-abi/.buildstamp] Error 1
make[1]: *** Waiting for unfinished jobs....
```

mhdawson commented:
Seems like there were places the function was still being used as the error was:

make[2]: Entering directory `/home/iojs/build/workspace/node-test-commit-linux-abi/nodes/ubuntu1404-64/test/addons/02_object_factory/build'
  CXX(target) Release/obj.target/addon/addon.o
../binding.cc: In function 'void CreateFunction(napi_env, napi_callback_info)':
../binding.cc:26:48: error: 'napi_set_function_name' was not declared in this scope
   status = napi_set_function_name(env, fn, name);
                                                ^

While looking at that can you also look at these warnings:
```bash
../binding.cc:35:63: warning: missing initializer for member 'napi_property_descriptor::getter' [-Wmissing-field-initializers]
   napi_property_descriptor desc = { "exports", CreateFunction };
                                                               ^
../binding.cc:35:63: warning: missing initializer for member 'napi_property_descriptor::setter' [-Wmissing-field-initializers]
../binding.cc:35:63: warning: missing initializer for member 'napi_property_descriptor::value' [-Wmissing-field-initializers]
../binding.cc:35:63: warning: missing initializer for member 'napi_property_descriptor::attributes' [-Wmissing-field-initializers]
../binding.cc:35:63: warning: missing initializer for member 'napi_property_descriptor::data' [-Wmissing-field-initializers]
```


We need to make sure we run ```make test``` on all PRs and validate that they pass.

boingoing commented:
Missed this one. I fixed the call to napi_create_function but forgot to remove the call to napi_set_function_name.

I thought the warnings reported for missing initializers were allowed by spec? Any missing initializers are converted to 0 as long as there was an initializer, I thought. It's just one line, so I went ahead and removed the warnings but I remember seeing this elsewhere.

I've been dev'ing this on Windows and found that ```vcbuild.cmd x64 debug test``` doesn't pass all tests. Should it?

boingoing committed:
Fix test build break and a couple warnings

boingoing commented:
Fixes https://github.com/nodejs/abi-stable-node/issues/78, by the way.

mhdawson commented:
I know all the tests should pass without the debug, is that clean ?  If so then I'd check the current Node master to see if the debug equivalent passes all tests or not.

mhdawson commented:
CI run: https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-commit-linux-abi/27/nodes=ubuntu1404-64/

mhdawson reviewed (approved):
Still LGTM after latest update.

jasongin commented on src/node_jsvmapi.h:
```diff
@@ -110,6 +110,7 @@ NODE_EXTERN napi_status napi_create_symbol(napi_env e,
 NODE_EXTERN napi_status napi_create_function(napi_env e,
                                              napi_callback cb,
                                              void* data,
+                                             napi_propertyname name,
```
I think this should be `const char*` instead of `napi_propertyname`. The function name is not used as a property key.

boingoing replied:
That's true and matches how ```napi_define_class``` names the constructor function. Inside napi, both ```napi_create_function``` and ```napi_define_class``` need to construct ```napi_value``` for the name string but this is an implementation detail. I'll change this to ```const char*``` as it's more consistent.

jasongin replied:
Maybe also put the name parameter second to be consistent with the parameter order of `napi_define_class` ?

boingoing replied:
Yeah, have to agree with you on that as well. Good point.

boingoing committed:
Change napi_create_function to take the name parameter as const char* instead of napi_propertyname

boingoing committed:
Move the utf8name parameter to napi_create_function into the second slot to match napi_define_class

jasongin reviewed (approved):
LGTM

mhdawson commented:
Think this broke the tests, see https://github.com/nodejs/abi-stable-node/issues/148
